tasks/test_results_processor.py

```diff
@@ -179,156 +180,219 @@ def _bulk_write_tests_to_db(
-                .values(test_flag_bridge_data)
-                .on_conflict_do_nothing(index_elements=["test_id", "flag_id"])
-            )
-            db_session.execute(insert_on_conflict_do_nothing_flags)
-            db_session.commit()
+        if len(test_flag_bridge_data) > 0:
+            self.save_test_flag_bridges(db_session, test_flag_bridge_data)
 
         log.info(
             "Inserted new test flag bridges to database",
             extra=dict(upload_id=upload_id),
         )
 
-        # Upsert Daily Test Totals
         if len(daily_totals) > 0:
-            rollup_table = DailyTestRollup.__table__
-            stmt = insert(rollup_table).values(
-                sorted(daily_totals.values(), key=lambda x: str(x["test_id"]))
-            )
-            stmt = stmt.on_conflict_do_update(
-                index_elements=[
-                    "repoid",
-                    "branch",
-                    "test_id",
-                    "date",
-                ],
-                set_={
-                    "last_duration_seconds": stmt.excluded.last_duration_seconds,
-                    "avg_duration_seconds": (
-                        rollup_table.c.avg_duration_seconds
-                        * (rollup_table.c.pass_count + rollup_table.c.fail_count)
-                        + stmt.excluded.avg_duration_seconds
-                    )
-                    / (rollup_table.c.pass_count + rollup_table.c.fail_count + 1),
-                    "latest_run": stmt.excluded.latest_run,
-                    "pass_count": rollup_table.c.pass_count + stmt.excluded.pass_count,
-                    "skip_count": rollup_table.c.skip_count + stmt.excluded.skip_count,
-                    "fail_count": rollup_table.c.fail_count + stmt.excluded.fail_count,
-                    "flaky_fail_count": rollup_table.c.flaky_fail_count
-                    + stmt.excluded.flaky_fail_count,
-                    "commits_where_fail": rollup_table.c.commits_where_fail
-                    + stmt.excluded.commits_where_fail,
-                },
+            sorted_rollups = sorted(
+                daily_totals.values(), key=lambda x: str(x["test_id"])
             )
-
-            db_session.execute(stmt)
-            db_session.commit()
+            self.save_daily_test_rollups(db_session, sorted_rollups)
 
         log.info(
             "Upserted daily test rollups to database", extra=dict(upload_id=upload_id)
         )
 
         # Save TestInstances
         if len(test_instance_data) > 0:
-            insert_test_instances = insert(TestInstance.__table__).values(
-                test_instance_data
-            )
-            db_session.execute(insert_test_instances)
-            db_session.commit()
+            self.save_test_instances(db_session, test_instance_data)
 
         log.info("Inserted test instances to database", extra=dict(upload_id=upload_id))
 
+    @sentry_sdk.trace
+    def create_daily_totals(
+        self,
```

Comment:
You probably don’t want to trace this either, as its a very simple function, and you call this thousands of times.

Another thing to note here is that you might want to make this a free function, as it does not depend on `self`.